### PR TITLE
Enhance usability of ModelZoo-YOLOX

### PR DESCRIPTION
**Pull request by illian01**
Slack link: https://nota-workspace.slack.com/archives/C040F65LSAJ/p1692766946904609
Notion link: https://www.notion.so/notaai/PyNP-ModelZoo-c43ab65458f54a78878d51535aa48e5d?pvs=4

목적
---
ModelZoo에서 세분화된 기능을 하나로 합쳐서 하나의 파일만 실행하면 경량화 → 재학습 → onnx export까지 되어 LaunchX 사용 가능하게 한다.

변경 사항
---
- auto_process.py 파일을 생성했습니다. fx 변환, 경량화, 재학습, onnx 변환 이라는 단계를 거치며, 이를 위해 다음과 같은 argument를 받습니다.
  - 모델의 이름 (yolox-s)
  - 모델 weight path
  - gpu device의 수
  - batch size
  - NetsPresso 이메일 계정
  - NetsPresso 계정 비밀번호
  - --fp16
- python auto_train_compress.py -n yolox-s -w yolox_s.pth -d 1 -b 64 --np_email {email} --np_password {password} --fp16
- 최대한 기존 코드의 수정을 막기 위해, 기존 코드들을 가져와 순서대로 작동하도록 구성했습니다.
- 중간에 발생하는 fx와 onnx 파일은 root 디렉토리에 생성되며, 재학습에서 발생하는 파일들은 YOLOX_outputs에 생성됩니다.

참조
---
- @shinkookchoi 신국님, @Jae-WoongYun 재웅님, 레포에 권한이 없어 reviewer 대신 태그로 참조 드립니다.

### CODE REVIEW

File: auto_process.py [hunk 1]
```diff
@@ -0,0 +1,340 @@
...
+        main,
+        num_gpu,
+        args.num_machines,
+        args.machine_rank,
+        backend=args.dist_backend,
+        dist_url=dist_url,
+        args=(exp, args),
+    )
+
+    logger.info("Fine-tining step end.")
```
Comment (Jae-WoongYun) [inline auto_process.py]: Fine-tining -> Fine-tuning 오타인거 같습니다.

File: auto_process.py [hunk 2]
```diff
@@ -0,0 +1,340 @@
...
+    )
+    parser.add_argument(
+        "--machine_rank", default=0, type=int, help="node rank for multi-node training"
+    )
+    parser.add_argument(
+        "--fp16",
+        dest="fp16",
+        default=False,
+        action="store_true",
+        help="Adopting mix precision training.",
```
Comment (Jae-WoongYun) [inline auto_process.py]: mix -> mixed로 수정 부탁드려요

File: auto_process.py [hunk 3]
```diff
@@ -0,0 +1,340 @@
...
+        task=TASK,
+        framework=FRAMEWORK,
+        file_path=UPLOAD_MODEL_PATH,
+        input_shapes=INPUT_SHAPES,
+    )
+
+    COMPRESSED_MODEL_NAME = "test_l2norm"
+    COMPRESSION_METHOD = CompressionMethod.PR_L2
+    RECOMMENDATION_METHOD = RecommendationMethod.SLAMP
+    RECOMMENDATION_RATIO = 0.6
```
Comment (Jae-WoongYun) [inline auto_process.py]: 개인적인 생각이지만, 사용자가 편하게 쓰는 목적으로 수정없이 쓴다면, 시작점으로 0.6은 좀 높은것 같습니다.
best practice 확인해보니 0.4일때 mAP -1.06에 latency 1.44x 인데 오타 수정하시면서 0.4로 해놓는게 어떨까요? 
(웹이랑 docs에선 0.5로 되어있긴하네요)
참고로 0.7일때 mAP -5.47, latency 2.89x 인데, 고민해보시고 오타 수정하시면서 같이 수정 부탁드릴게요.

File: auto_process.py [hunk 4]
```diff
@@ -0,0 +1,340 @@
...
+        "--np_email",
+        help="NetsPresso login e-mail",
+        type=str,
+    )
+    parser.add_argument(
+        "-p",
+        "--np_password",
+        help="NetsPresso login password",
+        type=str,
+    )
```
Comment (shinkookchoi) [inline auto_process.py]: `Compression arguments`에 compression method에 대한 것을 추가하면 어떨까요? 아래 값들에 대해서 입력할 수 있게 해주고, 기본값을 선택하면 좋을 것 같습니다.
```
COMPRESSION_METHOD = CompressionMethod.PR_L2
RECOMMENDATION_METHOD = RecommendationMethod.SLAMP
RECOMMENDATION_RATIO = 0.6
```

File: auto_process.py [hunk 5]
```diff
@@ -0,0 +1,340 @@
...
+    INPUT_SHAPES = [{"batch": 1, "channel": 3, "dimension": exp.input_size}]
+    model = compressor.upload_model(
+        model_name=UPLOAD_MODEL_NAME,
+        task=TASK,
+        framework=FRAMEWORK,
+        file_path=UPLOAD_MODEL_PATH,
+        input_shapes=INPUT_SHAPES,
+    )
+
+    COMPRESSED_MODEL_NAME = "test_l2norm"
```
Comment (shinkookchoi) [inline auto_process.py]: compressed model name을 자동으로 compression method에 따라 정할수 있을까요?
예를들어 `{original_model_name}_{compression_method}_{ratio} `이런식으로 자동 설정해주면 좋을 것 같습니다.

File: auto_process.py [hunk 6]
```diff
@@ -0,0 +1,340 @@
...
+        Implemented loggers include `tensorboard` and `wandb`.",
+        default="tensorboard"
+    )
+
+    """
+        Export arguments
+    """
+    parser.add_argument(
+        "--onnx_output_name", type=str, default="yolox.onnx", help="output name of models"
+    )
```
Comment (shinkookchoi) [inline auto_process.py]: 최종적으로 출력파일의 이름이라면 `COMPRESSED_MODEL_NAME`과 통일시키는게 어떨까요?

File: auto_process.py [hunk 7]
```diff
@@ -0,0 +1,340 @@
...
+        dist_url=dist_url,
+        args=(exp, args),
+    )
+
+    logger.info("Fine-tining step end.")
+
+    """ 
+        Export YOLOX model to onnx
+    """
+    logger.info("Export step start.")
```
Comment (shinkookchoi) [inline auto_process.py]: logger start와 end에서
onnx format으로 export한다고 명시적으로 표현해주면 좋을 것 같습니다.

**Comment by shinkookchoi** [comment]
아래의 순서로 잘 구성해주셨는데요. 각 repo마다의 형태는 다르겠지만 아래 순서대로 통일하면 좋을 것 같습니다. 
1. Convert YOLOX model to fx
2. Model compression - recommendation compression 
3. Retrain YOLOX model 
4. Export YOLOX model to onnx

File: auto_process.py [hunk 8]
```diff
@@ -233,11 +247,11 @@ def main(exp: Exp, args):
         input_shapes=INPUT_SHAPES,
     )
 
-    COMPRESSED_MODEL_NAME = "test_l2norm"
-    COMPRESSION_METHOD = CompressionMethod.PR_L2
-    RECOMMENDATION_METHOD = RecommendationMethod.SLAMP
-    RECOMMENDATION_RATIO = 0.6
-    OUTPUT_PATH = exp.exp_name + '_compressed.pt'
+    COMPRESSION_METHOD = args.compression_method
+    RECOMMENDATION_METHOD = args.recommendation_method
```
Comment (shinkookchoi) [inline auto_process.py:251]: `--compression_method`와 `--recommendation_method`로 입력 받는게 그대로 들어가는 건가요?
```
COMPRESSION_METHOD = CompressionMethod.PR_L2
RECOMMENDATION_METHOD = RecommendationMethod.SLAMP
```
argument들이 PR_L2등의 string을 입력받게 되는데요. 그렇게 하더라도 위의 예시와 동일하게 동작하나요? PyNP쪽 문법 확인하셔서 작성한 것인지 확인차 여쭙니다.

Comment (illian01) [reply auto_process.py:251]: @shinkookchoi 네. 동일하게 동작합니다. PyNP 에서 CompressionMethod 클래스가 아래와 같이 정의되어 있습니다. CompressionMethod.PR_L2 는 실제로 문자열 "PR_L2"와 같습니다. 이는 RecommendationMethod 도 유사하게 구성되어 있습니다.
argument를 입력으로 받은 뒤 다시 Enum에 접근해 문자열 값을 가져오는 대신, 정해진 문자열을 argument로 입력 하도록 하는 것이 구현적으로 차이가 없다고 생각했습니다. 이를 위해서, compression_method와 recommendation_method 두 argument는 choices를 추가해 입력 값에 제한을 두었습니다.

```
class CompressionMethod(str, Enum):
    PR_L2 = "PR_L2"
    PR_GM = "PR_GM"
    PR_NN = "PR_NN"
    PR_ID = "PR_ID"
    FD_TK = "FD_TK"
    FD_CP = "FD_CP"
    FD_SVD = "FD_SVD"

    @classmethod
    def create_literal(cls):
        return Literal["PR_L2", "PR_GM", "PR_NN", "PR_ID", "FD_TK", "FD_CP", "FD_SVD"]
```

Comment (shinkookchoi) [reply auto_process.py:251]: 설명 감사합니다.

**Review by shinkookchoi — APPROVED** [review]

File: auto_process.py [hunk 9]
```diff
@@ -0,0 +1,354 @@
...
+        recommendation_method=RECOMMENDATION_METHOD,
+        recommendation_ratio=RECOMMENDATION_RATIO,
+        output_path=OUTPUT_PATH,
+    )
+
+    logger.info("Compression step end.") 
+    
+    """ 
+        Retrain YOLOX model 
+    """
```
Comment (shinkookchoi) [inline auto_process.py:268]: compress 이후 retrain할때 train시 사용한 learning rate 값의 1/10 한 값을 기본값으로 넣으면 좋겠습니다.

Comment (illian01) [reply auto_process.py:268]: @shinkookchoi 
retrain 전에 exp 에 접근해 lr을 줄이는 방법으로 코드를 한줄 추가했습니다.
yolox 외에 다른 모델들도 추가해두는게 좋을까요?

Comment (shinkookchoi) [reply auto_process.py:268]: @illian01 네 경량화 이후 재학습하는 경우에는 원래 학습할때의 lr*0.1로 통일시켜서 값을 넣어주도록 추가해주세요.

**Review by Jae-WoongYun — APPROVED** [review]
